fixed the rock-paper-scissors docs example

alexhroom opened the pull request:
Fixes a mistake in the doc page added by #1413 - it turns out that strategies *can* recognise actions from different games corresponding to the same row/column, so long as both of the `Action` classes have a total ordering. That is:

Using the example code from before this PR:
`axl.rockpaperscissors.RPSAction.P == axl.Action.D` -> `False`
After this PR:
`axl.rockpaperscissors.RPSAction.P == axl.Action.D` -> `True`

This is good news for the interoperability of different games! :)

alexhroom committed:
fixed the RPS example

alexhroom committed:
fixed docs for doctest

marcharper commented on docs/tutorials/implement_new_games/index.rst:
```diff
@@ -53,12 +48,17 @@ has :code:`flip`, which flips a :code:`C` to a :code:`D` and vice versa!)
     ...     P = 1  # paper
     ...     S = 2  # scissors
     ...     
+    ...     def __lt__(self, other):
```
Why do we need this? Are we ever relying on R < P < S ? Also it's technically a circular ordering, since S < R.

alexhroom replied:
I mentioned a more full reason here: https://github.com/Axelrod-Python/Axelrod/pull/1418#issuecomment-1537077559 and also in the paragraph below this code block in the docs page itself.

but outside of interoperability between games (which we may or may not want) it's there for best practice (and to avoid user confusion if someone creates a game with a strategy containing a line like `if action == P`), and to keep consistency with the library itself [it has been implemented in exactly the same way as the IPD `Action` class implements ordering of actions](https://github.com/Axelrod-Python/Axelrod/blob/784cdd361bc40119b828e97801d0bb58db3a5669/axelrod/action.py#L21-L33). I can change it to just an `__eq__` method if you prefer and remove the `@total_ordering`.

drvinceknight commented:
Not actually sure what's causing the sphinx failure on read the docs. Here's the stack trace:

```
raceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/axelrod/envs/1419/lib/python3.7/site-packages/sphinx/registry.py", line 472, in load_extension
    mod = __import__(extname, None, None, ['setup'])
  File "/home/docs/checkouts/readthedocs.org/user_builds/axelrod/envs/1419/lib/python3.7/site-packages/sphinx/builders/linkcheck.py", line 19, in <module>
    from requests.exceptions import HTTPError
  File "/home/docs/checkouts/readthedocs.org/user_builds/axelrod/envs/1419/lib/python3.7/site-packages/requests/__init__.py", line 43, in <module>
    import urllib3
  File "/home/docs/checkouts/readthedocs.org/user_builds/axelrod/envs/1419/lib/python3.7/site-packages/urllib3/__init__.py", line 39, in <module>
    "urllib3 v2.0 only supports OpenSSL 1.1.1+, currently "
ImportError: urllib3 v2.0 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with OpenSSL 1.0.2n  7 Dec 2017. See: https://github.com/urllib3/urllib3/issues/2168

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/axelrod/envs/1419/lib/python3.7/site-packages/sphinx/cmd/build.py", line 303, in build_main
    args.tags, args.verbosity, args.jobs, args.keep_going)
  File "/home/docs/checkouts/readthedocs.org/user_builds/axelrod/envs/1419/lib/python3.7/site-packages/sphinx/application.py", line 224, in __init__
    self.setup_extension(extension)
  File "/home/docs/checkouts/readthedocs.org/user_builds/axelrod/envs/1419/lib/python3.7/site-packages/sphinx/application.py", line 449, in setup_extension
    self.registry.load_extension(self, extname)
  File "/home/docs/checkouts/readthedocs.org/user_builds/axelrod/envs/1419/lib/python3.7/site-packages/sphinx/registry.py", line 475, in load_extension
    raise ExtensionError(__('Could not import extension %s') % extname, err)
sphinx.errors.ExtensionError: Could not import extension sphinx.builders.linkcheck (exception: urllib3 v2.0 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with OpenSSL 1.0.2n  7 Dec 2017. See: https://github.com/urllib3/urllib3/issues/2168)

Extension error:
Could not import extension sphinx.builders.linkcheck (exception: urllib3 v2.0 only supports OpenSSL 1.
```

I believe this is fixed here: https://github.com/readthedocs/readthedocs.org/issues/10290. Will open a PR (hopefully) a fix to your branch @alexhroom.

drvinceknight committed:
Add a read the docs config file.

This is to fix a build problem that surfaced on https://github.com/Axelrod-Python/Axelrod/pull/1419

drvinceknight commented:
I've opened https://github.com/alexhroom/Axelrod/pull/1

alexhroom committed:
Merge pull request #1 from Axelrod-Python/add-readthe-docs-config-file

Add a read the docs config file.

alexhroom committed:
added numpy to doc requirements

alexhroom commented:
@drvinceknight looks like it's working - i also had to add numpy to the doc requirements.txt since it isn't being mocked.

marcharper commented:
> axl.rockpaperscissors.RPSAction.P == axl.Action.D -> True

I'm still not convinced we want this to be true -- it feels like a type mismatch. axl.action.D could reasonably equal RPSAction.S as well (being the last play rather than the second play). We're forcing a specific permutation of the actions for not a lot of benefit IMO.

Regardless of the mapping chosen, a 2x2 player like TFT may be able to technically play a game of higher dimension, but only in a degenerate way (restricted to a subset of actions). I don't see that this is really useful, other than perhaps refactoring some of the (degenerate) pure strategies like Cooperator and Defector to be more generic.

alexhroom commented:
@marcharper it matters a lot less now we've got a better idea on how we're going to redesign/refactor strategies - i'll leave this for now and create a separate PR just with the RTD fix.
